pkg/secret/ensure.go

```diff
@@ -46,3 +60,85 @@ func Ensure(client kubernetes.Interface, namespace string, secret *v1.Secret) (*
 
 	return object.(*v1.Secret), errors.Wrap(err, "error creating secret")
 }
+
+func EnsureFromSA(client kubernetes.Interface, saName, namespace string) (*v1.Secret, error) {
+	sa, err := client.CoreV1().ServiceAccounts(namespace).Get(context.TODO(), saName, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get ServiceAccount %s/%s", namespace, saName)
+	}
+
+	saSecret, err := getSecretFromSA(client, sa)
+
+	if err == nil {
+		return saSecret,nil
+	}
+
+	// We couldn't find right secret from this SA, search all Secrets
+	saSecret, err = getSecretForSA(client, sa)
+	if err != nil {
+		newSecret := &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-token-%s", sa.Name, generateRandomString(5)),
+				Namespace: namespace,
+				Annotations: map[string]string{
+					saAnnotation:        saName,
+					createdByAnnotation: creatorName,
+				},
+			},
+			Type: v1.SecretTypeServiceAccountToken,
+		}
+
+		saSecret, err = Ensure(client, newSecret.Namespace, newSecret)
+	}
+
+	secretRef := v1.ObjectReference{
+		Name: saSecret.Name,
+	}
+
+	sa.Secrets = append(sa.Secrets, secretRef)
+	_, err = serviceaccount.Ensure(client, namespace, sa)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to update ServiceAccount %v with Secret reference %v", saName, secretRef.Name)
+	}
+
+	return saSecret, err
+}
+
+func getSecretFromSA(client kubernetes.Interface, sa *v1.ServiceAccount) (*v1.Secret, error) {
+	secretNamePrefix := fmt.Sprintf("%s-token-", sa.Name)
+	for _, saSecretRef := range sa.Secrets {
+		if strings.HasPrefix(saSecretRef.Name, secretNamePrefix) {
+			saSecret, _ := client.CoreV1().Secrets(sa.Namespace).Get(context.TODO(), saSecretRef.Name, metav1.GetOptions{})
+			if saSecret.Annotations[saAnnotation] == sa.Name && saSecret.Type == v1.SecretTypeServiceAccountToken {
+				return saSecret, nil
+			}
+		}
+	}
+
+	return nil, errors.New(fmt.Sprintf("secret token for service-account %v/%v not found", sa.Namespace, sa.Name))
+}
+
+func getSecretForSA(client kubernetes.Interface, sa *v1.ServiceAccount) (*v1.Secret, error) {
+	saSecrets, err := client.CoreV1().Secrets(sa.Namespace).List(context.TODO(), metav1.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector("type", "kubernetes.io/service-account-token").String(),
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	for _, saSecret := range saSecrets.Items {
+		if saSecret.Annotations[saAnnotation] == sa.Name {
+			return &saSecret, nil
+		}
+	}
+
+	return nil, errors.New(fmt.Sprintf("secret for service-account %v/%v not found", sa.Namespace, sa.Name))
```

Comment:
We should return `NotFound` error so the caller can distinguish it.

```suggestion
	return nil, apierrors.NewNotFound(schema.GroupResource{
	    Group:    v1.SchemeGroupVersion.Group,
             Resource: "secrets",
        }, sa.Name)
```